File: backend/src/main/java/sepdrive/gruppen/backend/service/ChatService.java
package sepdrive.gruppen.backend.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import sepdrive.gruppen.backend.dto.ChatMessageRequest;
import sepdrive.gruppen.backend.dto.ChatMessageResponse;
import sepdrive.gruppen.backend.entity.ChatMessage;
import sepdrive.gruppen.backend.entity.Role;
import sepdrive.gruppen.backend.entity.User;
import sepdrive.gruppen.backend.repository.ChatMessageRepository;
import sepdrive.gruppen.backend.repository.UserRepository;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class ChatService {

    private final ChatMessageRepository chatMessageRepository;
    private final UserRepository userRepository;

    @Autowired
    public ChatService(ChatMessageRepository chatMessageRepository,
                       UserRepository userRepository) {
        this.chatMessageRepository = chatMessageRepository;
        this.userRepository = userRepository;
    }

    public ChatMessageResponse sendMessage(ChatMessageRequest request) {
        ChatMessage message = new ChatMessage();
        message.setDriveOfferId(request.getDriveOfferId());
        message.setSenderUsername(request.getSenderUsername());
        message.setContent(request.getContent());
        message.setCreatedAt(LocalDateTime.now());
        message.setEdited(false);
        message.setReadByKunde(false);
        message.setReadByFahrer(false);
        chatMessageRepository.save(message);
        return mapToResponse(message);
    }

    public List<ChatMessageResponse> getMessages(Long driveOfferId) {
        return chatMessageRepository.findAllByDriveOfferId(driveOfferId)
                .stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    public void markChatOpened(Long driveOfferId, String role) {
        List<ChatMessage> messages = chatMessageRepository.findAllByDriveOfferId(driveOfferId);

        for (ChatMessage message : messages) {
            if (!message.getSenderUsername().equalsIgnoreCase(role)) {
                if (role.equalsIgnoreCase("kunde")) {
                    message.setReadByKunde(true);
                } else if (role.equalsIgnoreCase("fahrer")) {
                    message.setReadByFahrer(true);
                }
            }
        }
        chatMessageRepository.saveAll(messages);
    }

    public void editMessage(Long messageId, String username, String newContent) {
        ChatMessage message = chatMessageRepository.findById(messageId)
                .orElseThrow(() -> new IllegalArgumentException("Message not found"));

        if (!message.getSenderUsername().equals(username)) {
            throw new IllegalArgumentException("Cannot edit messages from another user.");
        }

        if (isMessageLocked(message)) {
            throw new IllegalStateException("Editing is no longer allowed. Message has already been read.");
        }

        message.setContent(newContent);
        message.setEdited(true);
        chatMessageRepository.save(message);
    }

    public void deleteMessage(Long messageId, String username) {
        ChatMessage message = chatMessageRepository.findById(messageId)
                .orElseThrow(() -> new IllegalArgumentException("Message not found"));

        if (!message.getSenderUsername().equals(username)) {
            throw new IllegalArgumentException("Cannot delete messages from another user.");
        }

        if (isMessageLocked(message)) {
            throw new IllegalStateException("Deleting is no longer allowed. Message has already been read.");
        }

        chatMessageRepository.delete(message);
    }

    private boolean isMessageLocked(ChatMessage message) {
        User user = userRepository.findByUsername(message.getSenderUsername())
                .orElseThrow(() -> new IllegalArgumentException("User not found: " + message.getSenderUsername()));

        Role role = user.getRole();

        if (role == Role.KUNDE) {
            return message.isReadByFahrer();
        } else if (role == Role.FAHRER) {
            return message.isReadByKunde();
        }

        return false;
    }

    private ChatMessageResponse mapToResponse(ChatMessage message) {
        ChatMessageResponse dto = new ChatMessageResponse();
        dto.setId(message.getId());
        dto.setSenderUsername(message.getSenderUsername());
        dto.setContent(message.getContent());
        dto.setCreatedAt(message.getCreatedAt());
        dto.setEdited(message.isEdited());
        return dto;
    }
}

File: frontend/src/services/ChatService.js
import axios from 'axios';

const API_BASE_URL =
    window.location.hostname === 'localhost'
        ? 'http://localhost:8080'
        : 'http://backend:8080';

const ChatService = {


    getMessages: async (driveOfferId) => {
        try {
            const response = await axios.get(`${API_BASE_URL}/api/chat/${driveOfferId}`);
            return response.data;
        } catch (error) {
            console.error('Error fetching data:', error.response?.data || error.message);
            throw error;
        }
    },


    markChatOpened: async (driveOfferId, role) => {
        try {
            const response = await axios.post(`${API_BASE_URL}/api/chat/${driveOfferId}/opened?role=${role}`);
            return response.data;
        } catch (error) {
            console.error('Error marking Chat as opened:', error.response?.data || error.message);
            throw error;
        }
    },

    editMessage: async (messageId, username, newContent) => {
        try {
            const response = await axios.put(`${API_BASE_URL}/api/chat/${messageId}?username=${username}&newContent=${newContent}`);
            return response.data;
        } catch (error) {
            console.error('Error editing message:', error.response?.data || error.message);
            throw error;
        }
    },

    deleteMessage: async (messageId, username) => {
        try {
            const response = await axios.delete(`${API_BASE_URL}/api/chat/${messageId}?username=${username}`);
            return response.data;
        } catch (error) {
            console.error('Error deleting Message:', error.response?.data || error.message);
            throw error;
        }
    }

};

export default ChatService;
